drivers/led/sn32/rgb_matrix_sn32f24xb.c

```diff
@@ -0,0 +1,424 @@
+}
+
+static void shared_matrix_rgb_disable_pwm(void) {
+    // Disable PWM outputs on column pins
+    for (uint8_t y = 0; y < LED_MATRIX_COLS; y++) {
+        pwmDisableChannelI(&PWMD1, chan_col_order[y]);
+    }
+}
+
+static void shared_matrix_rgb_disable_leds(void) {
+    // Disable LED outputs on RGB channel pins
+    for (uint8_t x = 0; x < LED_MATRIX_ROWS_HW; x++) {
+        writePinLow(led_row_pins[x]);
+    }
+}
+
+static void update_pwm_channels(PWMDriver *pwmp) {
+    matrix_row_t row_shifter = MATRIX_ROW_SHIFTER;
+    for (uint8_t col_idx = 0; col_idx < LED_MATRIX_COLS; col_idx++, row_shifter <<= 1) {
+#ifdef MATRIX_NO_SCAN
+#    if (DIODE_DIRECTION == ROW2COL)
+        // Scan the key matrix column
+        if (!matrix_scanned) {
+            matrix_locked = true;
+            matrix_read_rows_on_col(shared_matrix, col_idx, row_shifter);
+        }
+#    endif // DIODE_DIRECTION == ROW2COL
+#endif     // MATRIX_NO_SCAN
+        uint8_t led_index = g_led_config.matrix_co[row_idx][col_idx];
+        // Check if we need to enable RGB output
+        if (led_state[led_index].b != 0) enable_pwm |= true;
+        if (led_state[led_index].g != 0) enable_pwm |= true;
+        if (led_state[led_index].r != 0) enable_pwm |= true;
+        // Update matching RGB channel PWM configuration
+        switch (current_row % LED_MATRIX_ROW_CHANNELS) {
+            case 0:
+                if (enable_pwm) pwmEnableChannelI(pwmp, chan_col_order[col_idx], led_state[led_index].b);
+                break;
+            case 1:
+                if (enable_pwm) pwmEnableChannelI(pwmp, chan_col_order[col_idx], led_state[led_index].g);
+                break;
+            case 2:
+                if (enable_pwm) pwmEnableChannelI(pwmp, chan_col_order[col_idx], led_state[led_index].r);
+                break;
+            default:;
+        }
+    }
+}
+static void rgb_callback(PWMDriver *pwmp) {
+    chSysLockFromISR();
+    // Disable the interrupt
+    pwmDisablePeriodicNotificationI(pwmp);
+    // Advance to the next LED RGB channels
+    current_row++;
+    if (current_row >= LED_MATRIX_ROWS_HW) current_row = 0;
+    uint8_t last_row_idx = row_idx;
+    // Advance to the next key matrix row
+    if (current_row % LED_MATRIX_ROW_CHANNELS == 2) row_idx++;
+    if (row_idx >= LED_MATRIX_ROWS) row_idx = 0;
+    // Disable LED output before scanning the key matrix
+    shared_matrix_rgb_disable_leds();
+    shared_matrix_rgb_disable_pwm();
+#ifdef MATRIX_NO_SCAN
+#    if (DIODE_DIRECTION == COL2ROW)
+    // Scan the key matrix row
+    if (!matrix_scanned) {
+        matrix_locked = true;
+        matrix_read_cols_on_row(shared_matrix, row_idx);
+    }
+#    endif // DIODE_DIRECTION == COL2ROW
+#endif     // MATRIX_NO_SCAN
+    update_pwm_channels(pwmp);
+#ifdef MATRIX_NO_SCAN
+    // Assume we have finished scanning the matrix
+    if (last_row_idx > row_idx) {
+        matrix_scanned = true;
+        matrix_locked  = false;
+    }
+#endif // MATRIX_NO_SCAN
```

Comment:
The logic here does not seem to be correct — `matrix_scanned` is set after `row_idx` has looped back to 0, but `matrix_scan_custom()` may clear `matrix_scanned` at any time after that, and `row_idx` may be pointing to any row at that time. So it is possible that only some rows have been scanned instead of the whole matrix.

Here is one possible implementation that starts scanning from whatever row was current at the time `matrix_scanned` was reset, and stop scanning when the same row has been reached again:
```suggestion
#ifdef MATRIX_NO_SCAN
#    if (DIODE_DIRECTION == COL2ROW)
    // Scan the key matrix row
    static uint8_t first_scanned_row;
    if (!matrix_scanned) {
        if (!matrix_locked) {
            matrix_locked = true;
            first_scanned_row = row_idx;
        } else {
            if ((last_row_idx != row_idx) && (row_idx == first_scanned_row)) {
                matrix_locked = false;
                matrix_scanned = true;
            }
        }
        if (matrix_locked) {
            matrix_read_cols_on_row(shared_matrix, row_idx);
        }
    }
#    endif // DIODE_DIRECTION == COL2ROW
#endif     // MATRIX_NO_SCAN
    update_pwm_channels(pwmp);
```
(The `ROW2COL` case is ignored here — it would need to be handled separately.)